compiler/rustc_mir_transform/src/lower_intrinsics.rs

```diff
@@ -149,6 +149,29 @@ impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
                             terminator.kind = TerminatorKind::Goto { target };
                         }
                     }
+                    sym::read_via_copy => {
+                        if args.len() != 1 {
+                            span_bug!(terminator.source_info.span, "Wrong number of arguments");
+                        };
+                        let arg = args.pop().unwrap();
+                        let (Operand::Copy(place) | Operand::Move(place)) = arg else {
+                            span_bug!(terminator.source_info.span, "`Operand::Constant` not supported");
+                        };
+                        let derefed_place = place.project_deeper(&[ProjectionElem::Deref], tcx);
```

Comment:
Intrinsic lowering doesn't happen until runtime mir. However, beginning in [`AnalysisPhase::PostCleanup`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/syntax/enum.AnalysisPhase.html#variant.PostCleanup), derefs must be the first projection in a place. If you add a mir opt test that looks like this it should ICE:

```rust
struct Foo(*const u8);

read_via_copy(foo.0);
```

You can fix this either by introducing a temporary or just add to the documentation that the argument must be a local and cannot have any additional projections.

Reply:
Updated the doc-comment for the intrinsic, and enforced it in the lowering.